Refactor folder structure + consolidate README's

rasmuscnielsen opened the pull request.

rasmuscnielsen commented:
@Dhaiwat10 Perhaps you can help me implement this new folder structure.

Once this is merged, we'll need to update the Vercel deployment to point to the new directory.

Furthermore it'd be great if we could also 

- Deploy the API to Vercel
- Setup a new staging environment where we can test the contract on Mumbai.

Here's a breakdown with the steps as I see them.


### Merging this PR

1. Merge this PR
2. Vercel: Edit `pixel-avatars` root directory from `web-previewer` --> `web-client`


### Deploy the API to Vercel

1. Create a new project in Vercel called `pixel-avatars-api` of type `Other`. It should point to `web-server` root directory. Leave the build settings fields empty.
2. Nevermind too much about environment variables for now as we'll need to update it for production later on anyway.


### Setting up a staging environment

1. Create a new `staging` branch from the `main` branch.
2. I think: Go to Vercel --> pixel-avatars --> Settings --> Domains and add a new `pixel-avatars-staging.vercel.app` that uses `staging` branch.
3.  I think: Go to Vercel --> pixel-avatars-api --> Settings --> Domains and add a new `pixel-avatars-api-staging.vercel.app` that uses `staging` branch.
4. Fill environment vars for pixel-avatars project. Make them specific to preview (I guess that means any branch except `main`?). Use variables from `web-client/.env.example` in this PR. Leave `VUE_APP_INFURA_ID` as already is in Vercel. Set `VUE_APP_SERVER_URL` to `https://pixel-avatars-api-staging.vercel.app`
5. Fill environment vars for pixel-avatars-api project. Make them specific to preview. Set `GENESIS_RPC=wss://rinkeby.infura.io/ws/v3/{INFURA_ID_HERE}`. Set `GENESIS_TOKEN=0x25ed58c027921E14D86380eA2646E3a1B5C55A8b`. Set `PRIVATE_KEY` to the key I'll DM you :-) 

I guess for this to work we'll also need the DAO to create the two new subdomains and point them to Vercel. Any idea who controls DNS for DeveloperDao.com? 

Ideally I'd also have liked that the web-client uses the API from the exact same version as itself. However that  currently doesn't seem to be easily achievable with Vercel (https://github.com/vercel/vercel/discussions/5231), but hopefully there won't be changes to the API for know so it shouldn't be a huge problem. 

LMK in case of questions!

briangershon commented on contract/hardhat.config.ts:
```diff
@@ -0,0 +1,65 @@
+
+    return {
+        solidity: '0.8.9',
+        defaultNetwork: 'localhost',
```
In my testing PR, I changed this to `hardhat` so `npm test` runs without needing a test node running. Is it safe to change to `hardhat` as the default?

briangershon replied:
Or we can change `package.json` so the tests using `hardhat` network like this:

```
"test": "yarn prettier && hardhat test --network hardhat",
```

Do you have a preference?

briangershon replied:
If no preference, I'll go with the update to `package.json` and will copy over your new hardhat config into my test branch...

briangershon replied:
This is complete and updated in PR #58

rasmuscnielsen replied:
@briangershon Sounds like a good call. Not sure what the exact difference is between the two, so probably best this way 👍

briangershon replied:
My understanding is that `hardhat` basically runs the blockchain in-process so no need for blockchain node to be running separately. `localhost` expects a node to be running in a separate process/terminal and if not, the tests errors out.

briangershon commented:
- [ ] I noticed some lint warnings in ./contracts when running `yarn lint`

<img width="868" alt="Screen Shot 2021-12-29 at 8 08 44 PM" src="https://user-images.githubusercontent.com/139330/147721064-2ff5bd1e-4452-4eb9-bb5f-5691741062d3.png">

briangershon commented on contract/.solhint.json:
```diff
@@ -0,0 +1,7 @@
+{
```
I noticed that if `.solhint.json` is in `contracts` folder, my VSCode doesn't pick it up.

By moving this file into the root of the repository, it works for VSCode + `yarn lint` still works properly too.

rasmuscnielsen replied:
Thanks Brian! I've pushed your suggested fix for the prettier / linting issue 👍  Feel free to doublecheck the issue is resolved in your end now

briangershon replied:
Thanks, looks good!

briangershon reviewed (commented):
I ran through PR, mainly focusing on docs, linting, dev environment. Just had a few minor comments.

I didn't try to get the full system working locally.

Dhaiwat10 commented:
@rasmuscnielsen I added the new Pixel Avatars API project to Vercel: https://pixel-avatars-api-bice.vercel.app/. Also added the `.env` vars

I haven't done the domain config yet because it says the `staging` branch does not exist yet and we also don't know who controls the domain yet. I'll try to find that person

with-heart commented:
> 1. Create a new `staging` branch from the `main` branch.

We have to complete this step before we can complete the rest of the staging env setup. Vercel won't point a domain to a branch that doesn't exist yet!

> I guess for this to work we'll also need the DAO to create the two new subdomains and point them to Vercel. Any idea who controls DNS for DeveloperDao.com?

If I'm understanding this correctly, you want to configure the subdomains like this:

- `pixel-avatars-staging.developerdao.com` -> `pixel-avatars-staging.vercel.app`
- `pixel-avatars-api-staging.developerdao.com` -> `pixel-avatars-api-staging.vercel.app`

Is that correct?

with-heart commented:
Is the intent behind the staging changes for **all** deploys _except for production_ to use Mumbai?

rasmuscnielsen commented:
> @rasmuscnielsen I added the new Pixel Avatars API project to Vercel: https://pixel-avatars-api-bice.vercel.app/. Also added the `.env` vars
> 
> I haven't done the domain config yet because it says the `staging` branch does not exist yet and we also don't know who controls the domain yet. I'll try to find that person

Awesome, thanks! 

Feel free to merge this PR at your convenience, and update the Vercel root-directory afterwards. I'm thinking it's better you're the one to merge the PR (or someone else with Vercel access) so we can minimize downtime

rasmuscnielsen commented:
@with-heart 

> We have to complete this step before we can complete the rest of the staging env setup. Vercel won't point a domain to a branch that doesn't exist yet!

Yep that makes sense. As soon as this PR is merged we can create the `staging` branch from main :-) 

> If I understand this, you want to configure the subdomains like this:
> 
> * `pixel-avatars-staging.developerdao.com` -> `pixel-avatars-staging.vercel.app`
> * `pixel-avatars-api-staging.developerdao.com` -> `pixel-avatars-api-staging.vercel.app`
> 
> Is that correct?

Hmm that could be the case -  I'm not too sure what's the best way to achieve this. 

Basically I'd like that `pixel-avatars-staging.developerdao.com` at any time displays the `staging` branch version of the website (using preview `env` that runs on Mumbai).
So however the `pixel-avatars.developerdao.com` website is configured, it should be similar for the staging website only using the corresponding staging branch.

Hope that makes sense :-) 

> Is the intent behind the staging changes for all deploys except for production to use Mumbai?

- Yes. All previews (feature branches, PR's etc) should ideally use Mumbai and the `pixel-avatars-api-staging.developerdao.com` endpoint for api. 
- The intent for setting up `pixel-avatars-staging.developerdao.com` itself though is so we can have a fixed url that we can share around for a few beta testers and test the full functionality of minting at any time regardless of making a PR :-)

Dhaiwat10 commented:
@rasmuscnielsen I don't think I'm a good person to review the PR as I haven't been in touch with the code at all recently. If you and other team members think it's good, let's merge it!

briangershon commented:
Hey @rasmuscnielsen 

While playing with #31 I discovered `yarn install` doesn't work on `web-client` due to conflict. Vue and `eslint-config-prettier` require incompatible versions.

- [ ] The `eslint-config-prettier` version in `web-client/package.json` needs to be downgraded to `v6.0.0`. e.g. `"eslint-config-prettier": "^6.0.0"`

rasmuscnielsen commented:
@briangershon Hmm that's odd. On my local machine it works fine (even after deleting node_modules) although it does indeed give a warning. 

<img width="1512" alt="image" src="https://user-images.githubusercontent.com/8465957/147790842-6773dced-fb44-4ce9-a9a7-504c2c1d5a8f.png">

However I've followed your suggestion and downgraded which also removed the warning 👍 

Afterwards my editor (PhpStorm) lights up like a Christmas tree though 😄 Seemingly it suggests that ESLint wants 2 spaces after prettier enforces 4. 
Haven't tinkered around with it too much, it might just my editor acting up. Perhaps you can check in yours if anything needs tweaking?

briangershon commented:
I wonder if an older version of eslint defaults to 2 spaces instead of 4? The prettier configs are all 4 spaces.

I pulled down latest branch and see that `yarn lint` just reversed all the changes that `yarn prettier` had with the 2 spaces instead of 4. (`yarn lint` runs `vue-cli-service lint` which change the files).

Maybe we can create an `.eslintrc.js` that sets `vue-clie-service lint` back to 4 spaces? I tried a simple eslintrc but got a lot of errors. I'll keep playing with...

briangershon commented:
Or maybe we just change prettier to use only 2 spaces for web-client to match Vue's lint? That's probably best workaround.

briangershon commented:
Something is still off though with approach of just changing prettier config. If I run `yarn lint` about 10 files are changed, seems fine. But then `yarn prettier` (with the change to 4 spaces) changes about 30 files. So they both make different sets of changes.

We'll see who wins, Prettier or Lint! :-)

briangershon commented:
If I switch the dependency back, I see what you mean about `yarn` just showing warning, so just a minor problem in that case.

On mine, I was running `npm install` and install was dying with:

<img width="841" alt="Screen Shot 2021-12-30 at 4 04 02 PM" src="https://user-images.githubusercontent.com/139330/147795433-ee595d79-d56e-470b-aaba-a8d7b7c67221.png">

**Probably just makes sense going back to previous version, though will require devs to use `yarn`**

(which might be ok to standardize, since I think code base has a mix of `npm` and `yarn` users given various combinations of package-lock.json and yarn.lock files)

rasmuscnielsen commented:
I have now aligned eslint and prettier configuration so they are no longer overwriting each other unnecessarily. There is only a single edge case in the current code base that I couldn't align. 

@Dhaiwat10 Go ahead and merge this branch now at your convenience and proceed with the Vercel setup :-)

Dhaiwat10 commented:
@rasmuscnielsen would it be more convenient if we could get you access to Vercel? (You could take my seat as I don't need access to Vercel right now)

I don't want to slow down your work just because I am the one who access to Vercel for absolutely no reason

rasmuscnielsen commented:
@Dhaiwat10 For sure, that works too if that's alright with you :-) 

Who's the person to set that in motion?

Dhaiwat10 commented:
@rasmuscnielsen just DM me your email and I'll get you in!
